3.analyze_data/utils/well_significance_testing.py

```diff
@@ -0,0 +1,206 @@
+        # Store the other data associated with the results
+        for name, val in _comp_names.items():
+            _treatments[name].append(val)
+
+    return _treatments
+
+def filter_wells_by_cell_count(_df, _cutoff):
+    """
+    Parameters
+    ----------
+    _df: pandas.Dataframe
+        The dataframe which contains the wells, which will be potentially removed
+
+    _cutoff: Integer
+        The number of cells a well must have to remain in the dataframe
+
+    Returns
+    -------
+    _df: pandas.Dataframe
+        The filtered dataframe with only wells that have more cells than the cutoff
+    """
+
+    # Find the number of cells in each well
+    well_counts = _df["Metadata_Well"].value_counts()
+
+    # Find the wells that contain more cells than the cutoff
+    wells_kept = well_counts[well_counts > _cutoff].index
+
+    # Keep only the wells that contain more cells than the cutoff
+    _df = _df.loc[_df["Metadata_Well"].isin(wells_kept)]
+
+    return _df
+
+def strat_samp_wells(_welldf):
+    """
+    Parameters
+    ----------
+    _welldf: pandas.Dataframe
+        The dataframe which contains the wells, which will be stratify sampled by well
+
+    Returns
+    -------
+    The sampled cells stratified by well
+    """
+
+    # Find the number of cells in each well
+    total_cell_count = _welldf.shape[0]
+
+    def samp_well(_well_samp):
+        well_frac = _well_samp.shape[0] / total_cell_count
+        return _well_samp.sample(frac=well_frac, random_state=0)
+
+    return _welldf.groupby('Metadata_Well', group_keys=False).apply(samp_well)
+
+def get_treatment_comparison(_comp_functions, _treatment_paths, _probadf, _barcode_platemapdf, _control_cutoff = 50, _treat_cutoff = 50):
+    """
+    Parameters
+    ----------
+    _comp_functions: Dictionary of Dictionaries
+        The keys are the names of the statistical tests.
+        The keys of the subdictionaries are the following strings {statistical_test_function, comparison_metric}.
+        The statistical_test_function key contains a function as a value, which outputs a zipped result of the statistical test, which must be defined prior to execution.
+        The comparison_metric key contains the name of the comparison metric computed as a string
+
+    _treatment_paths: Dictionary of Dictionaries
+        The keys of this dictionary are strings of the treatment types {compound, crispr, orf}.
+        The dictionaries corresponding to each of these keys have the keys {metadata, platemap, treatment_column, Plate_Map_Name}.
+        The metadata dataframe can be accessed with the metadata key, the platemap dataframe can be accessed with the platemap key, and the name of the treatment column can be accessed with the treatment_column key.
+
+    _probadf: pandas.Dataframe
+        The predicted probabilities and associated metadata for each cell
+
+    _barcode_platemapdf: pandas.Dataframe
+        Maps the plate to the treatment type
+
+    _control_cutoff: Integer
```

Comment:
These have default values so maybe consider defining if these args are required to be passed and what the default is in the docstring.

Reply:
Added